README.md
# Bosses_Killed_lfr
A world of Warcraft addon that creates a row of buttons showing you how many bosses you've cleared in each LFR and the LFR buttons light up if you're currently queuing for the corresponding raid.  No more memorizing, or having to go through the dropdown menu and mouseover the small "!" icon to see which bosses are down and which you still need to kill all the time.  The original author was wT-

link: 
http://wow.curseforge.com/addons/bosseskilled-lfr/

http://www.curse.com/addons/wow/bosseskilled-lfr/


Changelog.txt
2015/02/25
- v1.7.1
- Fix button scale
- Corrected raiddata

2015/02/24
- v1.7
- Fixed boss order
- Update TOC for patch 6.1

2015/02/16
- v1.6 
- Removed codes that was for FlexRaid as it is no longer needed for WOD
- Adding WOD new raid info as of 6.0.3
- Opening code on github for everyone so u can help maintain this addon so it stays up to date.
https://github.com/raykai/Bosses_Killed_lfr

2015/02/15 
- v1.5 
- Picking up addon from wT, renaming addon from bossekilled to Bosses-Killed-lfr.

2013-12-11 
 wT  <wT>  Add the addon table to the global scope so any of the things mentioned
in the last update are actually possible (oops), plus added some comments and stuff


raid_data.lua
local _, addon = ...

-- Index is from the first argument of DungeonInfoFunc.
-- This table is needed as there's no way to find out how many encounters are in each part of the same raid (to my knowledge).
addon.raidData = {
	-- MoP Raid finder as of 5.4
	[527] = { numEncounters = 3, startFrom =  1}, -- Guardians of Mogu'shan
	[528] = { numEncounters = 3, startFrom =  4}, -- The Vault of Mysteries
	[529] = { numEncounters = 3, startFrom =  1}, -- The Dread Approach
	[530] = { numEncounters = 3, startFrom =  4}, -- Nightmare of Shek'zeer
	[526] = { numEncounters = 4, startFrom =  1}, -- Terrace of Endless Spring
	[610] = { numEncounters = 3, startFrom =  1}, -- Last Stand of the Zandalari
	[611] = { numEncounters = 3, startFrom =  4}, -- Forgotten Depths
	[612] = { numEncounters = 3, startFrom =  7}, -- Halls of Flesh-Shaping
	[613] = { numEncounters = 3, startFrom = 10}, -- Pinnacle of Storms
	[716] = { numEncounters = 4, startFrom =  1}, -- Vale of Eternal Sorrows
	[717] = { numEncounters = 4, startFrom =  5}, -- Gates of Retribution
	[724] = { numEncounters = 3, startFrom =  9}, -- The Underhold
	[725] = { numEncounters = 3, startFrom = 12}, -- Downfall
	-- MoP Flex raids as of 5.4
	[726] = { numEncounters = 4, startFrom =  1}, -- Vale of Eternal Sorrows
	[728] = { numEncounters = 4, startFrom =  5}, -- Gates of Retribution
	[729] = { numEncounters = 3, startFrom =  9}, -- The Underhold
	[730] = { numEncounters = 3, startFrom = 12}, -- Downfall
	--WoD Raid finder info as 6.0.3
	-- Highmaul LFR
    [849] = { numEncounters = 3, startFrom =  1}, -- Walled City
	[850] = { numEncounters = 3, startFrom =  4}, -- Arcane Sanctum
	[851] = { numEncounters = 1, startFrom =  7}, -- Imperator's Rise
    -- Blackfrock Foundry LFR
    [847] = { numEncounters = 3, startFrom =  1}, -- Slagworks
	[846] = { numEncounters = 3, startFrom =  4}, -- The Black Forge
	[848] = { numEncounters = 3, startFrom =  7}, -- Iron Assembly
	[823] = { numEncounters = 1, startFrom =  10}, -- Blackhand's Crucible
	-- WoD Raid finder info as 6.1.0
	-- Waiting for the info
}


main.lua
--[[
	The look of the buttons/numbers/arrow can be altered by adding this addon to the
	OptionalDeps of an external addon which then overrides the functions marked below.
]]


local _, addon = ...

BossesKilled = addon

function addon:PLAYER_LOGIN()

	-- Hook raid finder frame's onshow
	local function myRaidFinderFrame_OnShow(frame)
		self:CreateButtons(frame, GetNumRFDungeons, GetRFDungeonInfo, RaidFinderQueueFrame_SetRaid)
		self:UpdateButtonsAndTooltips(frame)
		self:UpdateQueueStatuses()
		self.UpdateArrow()
		self:RegisterEvent("LFG_UPDATE")
	end
	RaidFinderQueueFrame:HookScript("OnShow", myRaidFinderFrame_OnShow)

	-- Hook raid finder frame's onhide
	local function myRaidFinderFrame_OnHide(frame)
		self:UnregisterEvent("LFG_UPDATE")
	end
	RaidFinderQueueFrame:HookScript("OnHide", myRaidFinderFrame_OnHide)

	hooksecurefunc("RaidFinderQueueFrame_SetRaid", self.UpdateArrow)

	-- Our event dispatcher calls functions like addon[eventName](...), so point it to the right update function
	self.LFG_UPDATE = self.UpdateQueueStatuses
end
addon:RegisterEvent("PLAYER_LOGIN")

-- Creates the buttons.
function addon:CreateButtons(parentFrame, DungeonAmountFunc, DungeonInfoFunc, SetDropdownMenuFunc)
	local scale = self:GetButtonScale(DungeonAmountFunc())

	if not parentFrame.BossesKilledButtons then
		parentFrame.BossesKilledButtons = {}
	end
	local buttons = parentFrame.BossesKilledButtons

	for i = 1, DungeonAmountFunc() do
		local id, name = DungeonInfoFunc(i)
		local isAvailable, isAvailableToPlayer = IsLFGDungeonJoinable(id)

		 -- Only make a button if there's data for it, and it hasn't been already made. This gets called multiple times so it updates correctly when you open up more raids
		if isAvailable and isAvailableToPlayer and not buttons[id] and self.raidData[id] then
			local button = self:CreateButton(parentFrame, scale)
			button.dungeonID = id
			button.dungeonName = name

			buttons[id] = button
			parentFrame.lastButton = button

			-- I just realised a CheckButton might already have it's own FontString, but uh... whatever.
			button.number = self:CreateNumber(button)

			button:SetScript("OnEnter", function(this)
				if this.tooltip then
					self:ShowTooltip(this)
				end
			end)

			button:SetScript("OnClick", function(this)
				SetDropdownMenuFunc(this.dungeonID)
				self.UpdateArrow()

				-- This is to override the automatic highlighting when you click the button, while we want to use that to show queue status instead.
				-- I've no idea why simply overriding this OnClick and not doing a SetChecked doesn't disable the behavior.
				-- I probably shouldn't be using a CheckButton at all, but the SpellBookSkillLineTabTemplate looks pretty nice for the job.
				this:SetChecked(this.checked)
			end)
			button.checked = false
		end
	end
end

---------------------------------------- Overridables ----------------------------------------

-- Must return a CheckButton, or implement :SetChecked() and support OnClick scripts. Not 100% right now if anything else
function addon:CreateButton(parent, scale)
	local button = CreateFrame("CheckButton", parent:GetName().."BossesKilledButton"..tostring(id), parent, "SpellBookSkillLineTabTemplate")
	button:Show()

	if parent.lastButton then
		button:SetPoint("TOPLEFT", parent.lastButton, "BOTTOMLEFT", 0, -15)
	else
		-- I don't like this fix for SocialTabs. It looks ugly, but it's good enough until (if) I figure out a more aesthetically pleasing solution
		button:SetPoint("TOPLEFT", parent, "TOPRIGHT", (3 + (IsAddOnLoaded("SocialTabs") and ceil(32/scale) or 0)), -50)
	end

	button:SetScale(scale)
	button:SetWidth(32 + 16) -- Original width is 32

	-- Need to find the button's texture in the regions so we can resize it. I don't like this part, but I can't think of a better way in case it's not the first region returned. (Is it ever not?)
	for _, region in ipairs({button:GetRegions()}) do
		if type(region) ~= "userdata" and region.GetTexture and region:GetTexture() == "Interface\\SpellBook\\SpellBook-SkillLineTab" then
			region:SetWidth(64 + 24) -- Original width is 64
			break
		end
	end

	return button
end

function addon:GetButtonScale(numDungeons)
	-- Ok, I still don't understand anything about the positioning and sizing of stuff in WoW, but the target frame is about 280'ish tall and buttons 32 and who gives a shit about margins and aaargh I'm going crazy /headexplode
	-- Magic numbers! There's really no method to the madness, these numbers just happen to look ok
	-- Mite need min 480 still too see
	return min(480 / (numDungeons * 32), 1)
end

-- Must return a fontstring
function addon:CreateNumber(button)
	local number = button:CreateFontString(button:GetName().."Number", "OVERLAY", "SystemFont_Shadow_Huge3")
	number:SetPoint("TOPLEFT", -4, 4)
	number:SetPoint("BOTTOMRIGHT", 5, -5)
	return number
end

-- Called from :UpateArrow()
function addon:CreateArrow()
	local arrow = GroupFinderFrame:CreateTexture("BossesKilledArrow", "ARTWORK")
	arrow:SetTexture("Interface\\ChatFrame\\ChatFrameExpandArrow")
	arrow:SetTexCoord(1, 0, 0, 1) -- This somehow turns the arrow other way around. Magic. /shrug
	arrow:SetSize(32, 32) -- Originally 16, 16
	arrow:Hide()
	return arrow
end

--------------------------------------- Update functions -------------------------------------

function addon:UpdateButtonsAndTooltips(parentFrame)
	local buttons = parentFrame.BossesKilledButtons

	for id, button in pairs(buttons) do
		local numKilled = 0

		-- GetLFGDungeonNumEncounters returns the full set of bosses for the whole raid
		-- are split into several separate raids, so get the amount of bosses and where to start from from our data file
		local index = self.raidData[id].startFrom
		local numEncounters = self.raidData[id].numEncounters

		local tooltip = {{text = button.dungeonName}} -- Set up tooltip data with the dungeon name
		for i = index, numEncounters + index - 1 do

            if id == 847 and i == 3 then
				i = 7
			end

            if id == 846 and i == 4 then
                i = 3
            end
            if id == 846 and i == 6 then
                i = 8
            end

            if id == 848 and i == 7 then
                i = 4
            end
            if id == 848 and i == 8 then
                i = 6
            end

			local bossName, _, isKilled = GetLFGDungeonEncounterInfo(id, i)

			if isKilled then
				if ENABLE_COLORBLIND_MODE == "1" then
					table.insert(tooltip, {text = "Dead - "..bossName})
				else
					table.insert(tooltip, {text = bossName, color = {RED_FONT_COLOR.r, RED_FONT_COLOR.g, RED_FONT_COLOR.b}})
				end
				numKilled = numKilled + 1
			else
				if ENABLE_COLORBLIND_MODE == "1" then
					table.insert(tooltip, {text = "Alive - "..bossName})
				else
					table.insert(tooltip, {text = bossName, color = {GREEN_FONT_COLOR.r, GREEN_FONT_COLOR.g, GREEN_FONT_COLOR.b}})
				end
			end
		end
		button.tooltip = tooltip

		local statusString = numKilled.."/"..numEncounters
		if ENABLE_COLORBLIND_MODE == "1" then
			button.number:SetFormattedText("|c00ffffff%s|r", statusString)
		else
			button.number:SetFormattedText(self.textColorTable[statusString], statusString)
		end
	end
end

function addon:UpdateQueueStatuses()
	for id, button in pairs(RaidFinderQueueFrame.BossesKilledButtons) do
		local mode = GetLFGMode(LE_LFG_CATEGORY_RF, id);
		if mode == "queued" or mode == "listed" or mode == "rolecheck" or mode == "suspended" then
			button:SetChecked(true)
			button.checked = true -- This is for the PostClick script earlier
		else
			button:SetChecked(false)
			button.checked = false
		end
	end
end

-- Not a method 'cos it's used as callback for hooksecurefuncs so it would get the wrong "self"
function addon.UpdateArrow()
	local arrow
	if not addon.arrow then
		arrow = addon:CreateArrow()
		addon.arrow = arrow
	else
		arrow = addon.arrow
	end

	local parent
	if RaidFinderQueueFrame:IsVisible() then
		parent = RaidFinderQueueFrame
	else
		arrow:Hide()
		return
	end

	if parent.raid and parent.BossesKilledButtons[parent.raid] then
		local button = parent.BossesKilledButtons[parent.raid]
		arrow:SetParent(button) -- Re-set the parent so we inherit the scale, so our smaller LFR buttons get a smaller arrow
		arrow:SetPoint("LEFT", button, "RIGHT")
		arrow:Show()
	end
end

--------------------------------- Tooltip and colorization stuff -----------------------------

function addon:ShowTooltip(button)
	GameTooltip:SetOwner(button, "ANCHOR_RIGHT")
	for i = 1, #button.tooltip do
		tooltip = button.tooltip[i]
		if tooltip.color then
			GameTooltip:AddLine(tooltip.text, unpack(tooltip.color))
		else
			GameTooltip:AddLine(tooltip.text)
		end
	end
	GameTooltip:AddLine(" ")
	GameTooltip:AddLine("<Click to select this raid>")
	GameTooltip:AddLine(" ")
	if ENABLE_COLORBLIND_MODE == "1" then
		GameTooltip:AddLine("(Button lights up if you're queued for this raid)")
	else
		GameTooltip:AddLine("(|c00ff0000Red|r for killed bosses, |c0000ff00green|r for alive ones,")
		GameTooltip:AddLine(" button lights up if you're queued for this raid)")
	end
	GameTooltip:Show()
end

local select, modf, format, strmatch = select, math.modf, format, strmatch
-- gets passed a "x/y" STRING. No sanity checks so make sure the calling function feeds it the expected format.
-- the vararg gets passed color triples, eg. 0.0,1.0,0.0,  1.0,0.0,0.0 (green to red)
function addon:TextColorGradient(str_percent, ...)
	local num = select("#", ...)
	local low, high = strmatch(str_percent, "(%d+)/(%d+)")
	local percent = (low + 0) / (high + 0) -- implicit cast to number, cheaper than tonumber
	local r, g, b

	if percent >= 1 then
		r, g, b = select(num - 2, ...), select(num - 1, ...), select(num, ...)
		return format("|cff%02x%02x%02x%%s|r", r * 255, g * 255, b * 255)
	elseif percent <= 0 then
		r, g, b = ...
		return format("|cff%02x%02x%02x%%s|r", r * 255, g * 255, b * 255)
	end

	local segment, relperc = modf(percent * (num / 3 - 1))
	local r1, g1, b1, r2, g2, b2
	r1, g1, b1 = select((segment * 3) + 1, ...), select((segment * 3) + 2, ...), select((segment * 3) + 3, ...)
	r2, g2, b2 = select((segment * 3) + 4, ...), select((segment * 3) + 5, ...), select((segment * 3) + 6, ...)

	if not r2 or not g2 or not b2 then
		r, g, b = r1, g1, b1
	else
		r, g, b = r1 + (r2 - r1) * relperc,
		          g1 + (g2 - g1) * relperc,
		          b1 + (b2 - b1) * relperc
	end
	return format("|cff%02x%02x%02x%%s|r", r * 255, g * 255, b * 255)
end

-- Use a memoization table so each x/y colorstring is only computed once and then does a simple lookup
addon.textColorTable = setmetatable({}, {__index = function(t, k)
	local colorStr = addon:TextColorGradient(k, GREEN_FONT_COLOR.r, GREEN_FONT_COLOR.g, GREEN_FONT_COLOR.b, -- "From" color
	                                            RED_FONT_COLOR.r,   RED_FONT_COLOR.g,   RED_FONT_COLOR.b)   -- "To" color
	rawset(t, k, colorStr)
	return colorStr
end})
